### PR TITLE
Introduce json report generation

### PR DESCRIPTION
**Pull request by fernflower**
That is an ugly hack to make leapp preupgrade produce
a json report alongwith default text one.
Json report will be present in the same directory as the
text report.

### CODE REVIEW

File: leapp/cli/upgrade/__init__.py [hunk 1]
```diff
@@ -211,6 +212,12 @@ def preupgrade(args):
         report_info(get_config().get('report', 'path'), fail=True)
         sys.exit(1)
     report_info(get_config().get('report', 'path'))
+    # fetch all report messages as a list of dicts
+    messages = fetch_upgrade_report_raw(context, renderers=False)
+    report_json = get_config().get('report', 'path').replace('.txt', '.json')
+    # XXX FIXME use leapp way of outputting things?
+    with open(report_json, 'w+') as f:
+        f.write(json.dumps({'entries': messages}))
```
Comment (vinzenz) [inline leapp/cli/upgrade/__init__.py]: json.dump? I don’t see why you need to create a temporary string - I know WIP and POC etc 
Just saying:-)

File: leapp/utils/report.py [hunk 2]
```diff
@@ -20,34 +20,57 @@ def _flatten_dict(dict_):
...
+    :return: A list of upgrade messages of type "Report" withing the given context,
+             each message is a dict of raw data.
+    """
+    report_msgs = get_messages(names=['Report'], context=context_id)
+    if not report_msgs:
+        # return empty list to be consistent?
+        return
+
+    messages = []
+    for message in report_msgs:
```
Comment (vinzenz) [inline leapp/utils/report.py]:
```python
    for message in report_msgs:
        data = json.loads(message.get('message').get('data'))
        merged_data = {
            'topic': message.get('topic'),
            'phase': message.get('phase'),
            'audience': data['audience'],
            'title': data['title'],
            'flags': data['flags'],
            'severity': data['severity']
        }
        if renderers:
            merged_data['renderers'] = data['renderers']
        merged_data.update(_flatten_dict(json.loads(data['detail'])))
        messages.append(merged_data)
```

Comment (fernflower) [reply leapp/utils/report.py]: yeah, will refactor, thanks

File: leapp/cli/upgrade/__init__.py [hunk 3]
```diff
@@ -211,6 +212,11 @@ def preupgrade(args):
         report_info(get_config().get('report', 'path'), fail=True)
         sys.exit(1)
     report_info(get_config().get('report', 'path'))
+    # fetch all report messages as a list of dicts
+    messages = fetch_upgrade_report_raw(context, renderers=False)
+    report_json = get_config().get('report', 'path').replace('.txt', '.json')
```
Comment (pirat89) [inline leapp/cli/upgrade/__init__.py]: would about
```suggestion
    report_json = get_config().get('report', 'path_json')
```
(require update of 'config' file as well)

Comment (fernflower) [reply leapp/cli/upgrade/__init__.py]: Honestly I see no point in defining separate paths for txt and json report, defining a 'leapp report dir' should be enough control over output. 
But replace('.txt', '.json') can be improved

**Comment by drehak** [comment]
Looks good to me personally.

**Comment by drehak** [comment]
Before we merge, shouldn't the json report be mentioned in the output too? Or will we make a separate PR for that?

File: leapp/cli/upgrade/__init__.py [hunk 4]
```diff
@@ -207,10 +208,16 @@ def preupgrade(args):
 
     report_errors(workflow.errors)
 
+    report_txt, report_json = [os.path.join(get_config().get('report', 'dir'),
+                                            'leapp-report.{}'.format(f)) for f in ['txt', 'json']]
+    report_info(report_txt, fail=workflow.errors)
+    # fetch all report messages as a list of dicts
+    messages = fetch_upgrade_report_raw(context, renderers=False)
+    report_json = os.path.join(get_config().get('report', 'dir'), 'leapp-report.json')
```
Comment (pirat89) [inline leapp/cli/upgrade/__init__.py]: @fernflower  probably this line should be removed

**Comment by pirat89** [comment]
@fernflower Additionally, when we are doing the change, I would like to see change in leapp-repository as well. One of reasons to have the paths defined was to have just one source of true. See the [line](https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/actors/verifycheckresults/actor.py#L26). When you do change, I expect there is a way to get the correct path so it will not be hardcoded here. Or the actor will be modified so the path will not be needed.

**Comment by pirat89** [comment]
but still I am unhappy that path is again hardcoded (and not sure how to provide the info to actors). even the dir doesn't look to me like something that's needed as it is the same one used for logs. I think that better solution would be to provide paths `report_txt` and `report_json` as with that the info doesn't have to be hardcoded anywhere else.

**Comment by fernflower** [comment]
@pirat89 I get you, I created a bug/enhancement https://github.com/oamg/leapp/issues/531

**Pull request merged**